src/graph/GraphRunCtx.cpp

```diff
@@ -0,0 +1,211 @@
+		;
+
+	for(auto&& node : executionOrder) {
+		RGL_DEBUG("Enqueueing node: {}", *node);
+                NvtxRange rg {graphOrdinal, NVTX_COL_WORK, "Enqueue({})", node->getName()};
+                node->enqueueExec();
+                executionStatus.at(node).enqueued.store(true);
+                executionStatus.at(node).enqueued.notify_all();
+	}
+	RGL_DEBUG("Node enqueueing done");  // This also logs the time diff for the last one
+}
+catch (...)
+{
+	// Exception most likely happened in a Node, but might have happened around executionOrder loop.
+	// We still need to communicate that nodes 'executed' (even though some may not have a chance to start).
+	// If we didn't, we could hang client's thread in synchronizeNodeCPU() waiting for a Node that will never run.
+	for (auto&& [node, state] : executionStatus) {
+		if (state.enqueued.load(std::memory_order::relaxed)) {
+			continue;
+		}
+		state.exceptionPtr = std::current_exception();
+		state.enqueued.store(true);
+		state.enqueued.notify_all();
+	}
+}
+
+std::vector<std::shared_ptr<Node>> GraphRunCtx::findExecutionOrder(std::set<std::shared_ptr<Node>> nodes)
+{
+	// Nodes already present in this vector need to be executed after the ones that are added later.
+	std::vector<std::shared_ptr<Node>> reverseOrder {};
+	std::function<void(std::shared_ptr<Node>)> dfsRec = [&](std::shared_ptr<Node> current) {
+		nodes.erase(current);
+		for (auto it = current->getOutputs().rbegin(); it != current->getOutputs().rend(); ++it) {
+			const auto& outputNode = *it;
+			if (nodes.contains(outputNode)) {
+				dfsRec(outputNode);
+			}
+		}
+		reverseOrder.push_back(current);
+	};
+	// Get entry nodes (graphs may have many)
+	std::vector<Node::Ptr> entryNode;
+	for (auto&& node : nodes) {
+		if (node->getInputs().empty()) {
+			entryNode.push_back(node);
+		}
+	}
+	// Sort entry nodes by priority, ascending!
+	std::stable_sort(entryNode.begin(), entryNode.end(),
+	                 [](Node::Ptr lhs, Node::Ptr rhs)
+	                 { return lhs->getPriority() < rhs->getPriority(); });
+
+	// Iterate in ascending order of entry priorities
+	for (auto&& node : entryNode) {
+		dfsRec(node);
+	}
+
+	if (!nodes.empty()) {
+		// I'm pretty sure the rest of the code guarantees it will not happen, but just in case for now.
+		throw std::logic_error("bug: detected unreachable nodes");
+	}
+	return {reverseOrder.rbegin(), reverseOrder.rend()};
+}
+
+GraphRunCtx::~GraphRunCtx()
+{
+	// If GraphRunCtx is destroyed, we expect that thread was joined and stream was synced.
+
+	// Log error if stream has pending work.
+	cudaError_t status = cudaStreamQuery(stream->getHandle());
+	if (status == cudaErrorNotReady) {
+		RGL_WARN("~GraphRunCtx(): stream has pending work!");
+		status = cudaSuccess; // Ignore further checks.
+	}
+	CHECK_CUDA_NO_THROW(status);
+
+	if (maybeThread.has_value()) {
+		RGL_WARN("~GraphRunCtx(): maybeThread has value!");
+	}
+}
+
+void GraphRunCtx::detachAndDestroy()
```

Comment:
This method is supposed to synchronize `GraphRunCtx` and remove all its references, which is expected to be followed by its destruction. After `detachAndDestroy`,  none of `this->nodes`  have `Node::graphRunCtx` set, so `Node::validate()` (ensuring `Node::graphRunCtx` is set) would fail with an exception.

Reply:
Clear, thank you!